test/offline/framework/orchestration.cpp

```diff
@@ -0,0 +1,289 @@
+    }
+    for (auto& file : coll.art_files) {
+      object_files.insert(file.unique_name);
+    }
+  }
+
+  dc.config->allowed_objects = object_files;
+  // don't try to do this because we can't write the file
+  dc.config->generate_symbol_definition_map = false;
+
+  std::vector<fs::path> dgo_paths;
+  for (auto& x : offline_config.dgos) {
+    dgo_paths.push_back(iso_data_path / x);
+  }
+
+  dc.db = std::make_unique<decompiler::ObjectFileDB>(dgo_paths, dc.config->obj_file_name_map_file,
+                                                     std::vector<fs::path>{},
+                                                     std::vector<fs::path>{}, *dc.config);
+
+  std::unordered_set<std::string> db_files;
+  for (auto& files_by_name : dc.db->obj_files_by_name) {
+    for (auto& f : files_by_name.second) {
+      db_files.insert(f.to_unique_name());
+    }
+  }
+
+  if (db_files.size() != object_files.size()) {
+    lg::error("DB file error: has {} entries, but expected {", db_files.size(),
+              object_files.size());
+    for (const auto& coll : work.work_collections) {
+      for (auto& file : coll.source_files) {
+        if (!db_files.count(file.unique_name)) {
+          lg::error(
+              "didn't find {}, make sure it's part of the DGO inputs and not in the banned objects "
+              "list\n",
+              file.unique_name);
+        }
+      }
+      for (auto& file : coll.art_files) {
+        if (!db_files.count(file.unique_name)) {
+          lg::error("didn't find {}\n", file.unique_name);
+        }
+      }
+    }
+    exit(1);
+  }
+
+  return dc;
+}
+
+std::vector<std::future<OfflineTestThreadResult>> distribute_work(
+    const OfflineTestConfig& offline_config,
+    const std::vector<OfflineTestSourceFile>& files,
+    const std::vector<OfflineTestArtFile>& art_files) {
+  // First, group files by their DGO so they can be partitioned
+  std::unordered_map<std::string, OfflineTestWorkCollection> work_colls = {};
+
+  for (const auto& file : files) {
+    if (work_colls.count(file.containing_dgo) == 0) {
+      work_colls[file.containing_dgo] = OfflineTestWorkCollection();
+    }
+    work_colls[file.containing_dgo].source_files.push_back(file);
+  }
+
+  for (const auto& file : art_files) {
+    if (work_colls.count(file.containing_dgo) == 0) {
+      work_colls[file.containing_dgo] = OfflineTestWorkCollection();
+    }
+    work_colls[file.containing_dgo].art_files.push_back(file);
+  }
+
+  // Now partition by DGO so that threads do not consume unnecessary or duplicate resources
+  // this is a half-decent approximation of a greedy-knapsack approach (where the knapsack can be
+  // overstuffed) Repeatedly just add to the thread with the current least amount of work
+  //
+  // TODO - if it ends up being that it would be advantageous to split up a massive dgo into
+  // multiple threads (ie. lots in engine) then this should be improved to accomodate that.
```

Comment:
another thing to call out, there's likely a good reason to duplicate `ENGINE.DGO` across a few threads, as it is a third of all files.